security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -195,90 +195,161 @@ func isIngressGatewaySecret(scrt *v1.Secret) bool {
 	return true
 }
 
-// extractCertAndKey extracts key, certificate and root certificate, and indicates whether
-// these key and certificate are empty.
-func extractCertAndKey(scrt *v1.Secret) (cert, key, root []byte, valid bool) {
+// extractCertAndKey extracts server key, certificate, and indicates whether key and cert exist.
+func extractCertAndKey(scrt *v1.Secret) (cert, key []byte, exist bool) {
 	certAndKeyExist := false
 	if len(scrt.Data[genericScrtCert]) > 0 {
 		cert = scrt.Data[genericScrtCert]
 		key = scrt.Data[genericScrtKey]
-		root = scrt.Data[genericScrtCaCert]
 	} else {
 		cert = scrt.Data[tlsScrtCert]
 		key = scrt.Data[tlsScrtKey]
-		root = []byte{}
 	}
-	// root could be empty if ingress gateway only accepts TLS.
 	if len(cert) > 0 && len(key) > 0 {
 		certAndKeyExist = true
 	}
-	return cert, key, root, certAndKeyExist
+
+	return cert, key, certAndKeyExist
 }
 
-func (sf *SecretFetcher) scrtAdded(obj interface{}) {
-	scrt, ok := obj.(*v1.Secret)
-	if !ok {
-		secretFetcherLog.Warnf("Failed to convert to secret object: %v", obj)
-		return
+// extractCACert extracts the client CA certificate from either the Compound
+// Secret, or from a separate Kubernetes TLS secret that has CA cert in `tls.crt` field.
+func extractCACert(scrt *v1.Secret, fromCompoundSecret bool) (caCert []byte, exist bool) {
+	if len(scrt.Data[genericScrtCaCert]) > 0 {
+		caCert = scrt.Data[genericScrtCaCert]
+	} else if !fromCompoundSecret {
+		caCert = scrt.Data[tlsScrtCert]
 	}
 
+	return caCert, len(caCert) > 0
+}
+
+// extractK8sSecretIntoSecretItem extracts a server cert/key pair and a client CA
+// certificate from the k8s Secret into a pair of SecretItems. Returns SecretItems and a boolean
+// indicating whether this is a CA only k8s Secret.
+// A CA only k8s secret has name suffix `-cacert`, and is ONLY considered for a client CA;
+// either a `cacert` or `tls.crt` must be provided.
+// Otherwise the Secret can hold a server cert/key pair in `tls.crt`/`tls.key`,
+// or a server cert/key pair in `cert`/`key` and an optional client CA cert in
+// `-cacert`. A Secret with server cert/key and client CA cert is considered as a compound secret.
+func extractK8sSecretIntoSecretItem(scrt *v1.Secret, t time.Time) (serverItem, clientCAItem *model.SecretItem, isCAOnlySecret bool) {
```

Comment:
I feel we could simply this func a bit by removing ```isCAOnlySecret``` as returned param, just return ```ns``` and ```cans```; caller logic will be simpler like 
```
if ns != nil {
  //update/push
}
if cans != nil {
  //update/push
}
```

Reply:
This is to differentiate client CA only secret and compound secret.
- If a compound secret `bookinfo` is created, but has invalid server cert. 
scrtAdded() is called, and extractK8sSecretIntoSecretItem() returns ns(nil), cans(not nil), isCaOnly(false). This secret will be discarded. Without returning and checking isCAOnlySecret, we will add this secret, which is not expected.
- If a CA only secret `bookinfo-cacert` is created, and extractK8sSecretIntoSecretItem() returns ns(nil), cans(not nil), isCaOnly(true). This secret will be added by scrtAdded().

Moreover,
- If a compound secret `bookinfo` is created, but has at least one invalid cert. extractK8sSecretIntoSecretItem() returns ns(nil), cans(nil), isCaOnly(false). This secret is discarded.
- If a secret `bookinfo` with only server key/cert is created, and cert if valid. This secret is added by scrtAdded(). Otherwise, it is discarded.

We don't want to load any secret that has invalid cert.